googlemock/docs/translations/ko_KR/for_dummies.md

```diff
@@ -0,0 +1,452 @@
+
+#### 정의하는 방법
+
+`Turtle` interface를 계속 사용한다. 기본적으로 mock class를 정의하기 위해서는 아래와 같은 단계를 거치게 된다.
+
+- `Turtle` interface를 상속받는 `MockTurtle` class를 생성한다.
+- `Turtle` interface에 정의한 *virtual* function의 이름과 해당 function이 몇 개의 argument를 가지고 있는지 확인한다. (상속 대신 템플릿을 이용하는 것도 가능하지만 상속보다는 약간 어렵다. 관련내용은 [여기](cook_book.md#nonvirtual-method-mocking-하기)에서 확인 가능하다.)
+- Derived class의 `public:` 영역에 `MOCK_METHOD();` macro를 사용한다.
+- 다음으로는 `Turtle` interface의 function signature에서 *function name*, *return type*, *argument*들을 복사해서 `MOCK_METHOD();`에 그대로 붙여 넣는다. (아래 예제코드 참조)
+- 만약, const method를 mocking하고 있다면 4번째 parameter에 `(const)`를 추가한다.
+- Virtual method를 overriding하고 있는 것이므로 `override` 키워드를 사용하기를 추천한다. 그렇게 되면 4번째 parameter는 `(override)` 혹은 `(const, override)`가 될 것이다.
+- 위의 내용을 `Turtle` interface의 다른 모든 virtual function에 대해서도 수행한다. C++에서 interface(abstract class)를 상속받는 derived class는 interface의 pure virtual method를 전부 override해야만 object를 생성할 수 있기 때문이다.
+
+위의 과정을 거치면 최종적으로 아래와 같은 모양이 된다.
+
+```cpp
+#include "gmock/gmock.h"  // Brings in gMock.
+
+class MockTurtle : public Turtle {
+ public:
+  ...
+  MOCK_METHOD(void, PenUp, (), (override));
+  MOCK_METHOD(void, PenDown, (), (override));
+  MOCK_METHOD(void, Forward, (int distance), (override));
+  MOCK_METHOD(void, Turn, (int degrees), (override));
+  MOCK_METHOD(void, GoTo, (int x, int y), (override));
+  MOCK_METHOD(int, GetX, (), (const, override));
+  MOCK_METHOD(int, GetY, (), (const, override));
+};
+```
+
+위와 같이 했다면 다른 추가작업은 필요 없다. 나머지는 `MOCK_METHOD` macro가 알아서 해주게 된다. 이렇듯 간단하게 mock class와 mock function을 정의할 수 있으며 조금만 익숙해지면 매우 빠르게 mock class를 생성할 수 있을 것이다.
+
+#### 어떤 파일에 저장해야 할까?
+
+Mock class를 정의할 때, 해당 class를 실제로 어디에 저장해야 할지도 고민해 볼 부분이다. 예를 들어 `Foo`라는 interface가 있다면 이것을 사용하는 모든 개발자가 본인만의 공간에 본인만의 테스트파일(예: `*_test.cc`)을 만들고 `MockFoo` class를 구현할 수 있다. 물론, 이렇게 해도 동작상에 문제는 없지만 유지보수 측면에서는 어려움이 발생할 것이다. 왜냐하면 `Foo` interface가 고쳐졌다고 가정하면, 독자적으로 이것을 상속받아 사용하던 모든 테스트가 실패하거나 영향을 받기 때문이다.
+
+따라서 되도록이면 interface와 mock class는 같은 패키지에 있는 것이 좋다. (이 경우에 production 빌드와 test 빌드를 확실히 구분해서 둘이 섞이지 않도록 주의해야 한다.) 즉, `Foo`가 포함된 패키지안에 `mock_foo.h`라는 파일을 만들고 `MockFoo`를 구현하면 된다. 그리고 모든 개발자가 `mock_foo.h`를 사용하게 한다. 이렇게 되면 `Foo`가 변경되더라도 1개의 `MockFoo`만 수정하면 되기 때문에 유지보수 측면에서 도움이 된다.
+
+또 다른 방법은 `Foo` interface보다 더 상위에 `FooAdaptor`라는 새로운 interface를 만드는 것이다. 이러한 패턴은 외부로 노출되는 `FooAdaptor`는 쉽게 변하지 않도록 유지하면서도 내부에서는 자유롭게 변경작업이 진행될 수 있도록 도와준다. 물론 초반에 해야할 일이 많아지는 것은 사실이지만 코드의 가독성을 높이고 구현난이도를 낮춰주는 장점이 있기 때문에 장기적으로 볼 때는 좋은 방법이라고 할 수 있다.
+
+### 테스트에서 Mock 이용하기
+
+Mock class를 생성한 이후에는 아래와 같은 방법으로 진행한다.
+
+- gMock 관련 기능을 사용하기 위해서 `testing` namespace를 포함시킨다. (한 파일에 한 번씩만)
+- Mock object를 생성한다.
+- Mock object에 expectation을 지정한다. (어떤 method를 호출할지, 어떤 argument를 전달받을지, 어떤 동작을 할지 등)
+- Mock object를 사용하는 실제코드를 수행한다. 이 때 expectation으로 지정된 횟수 및 argument 정보와 다르게 호출되는 부분이 있다면 즉시 에러로 검출된다. (결과확인을 위해 googletest assertion을 추가로 사용하는 것도 좋다.)
+- Mock object가 소멸될 때, gMock은 설정한 *모든* expectation들이 만족되었는지 확인해준다.
```

Comment:
Please use numbered list. The next paragraph and the sample code mention the numbers.

Reply:
Please check result here https://github.com/google/googletest/pull/2513/commits/2e9001784e73743363e34e347e89dbbe8def2322  
and let me know if there are some mistakes